src/collections/_documentation/development/sdk-dev/envelopes.md

```diff
@@ -0,0 +1,448 @@
+- Although not part of the Headers, Envelopes are always terminated by a newline
+  (`\n`) or the end of the file.
+- Empty headers `{}` are technically valid.
+
+Header-only Example:
+
+```json
+{"event_id":"12c2d058-d584-4270-9aa2-eca08bf20986"}
+```
+
+### Envelopes
+
+The full grammar for an Envelope is:
+
+```
+Envelope = Headers { "\n" Item } [ "\n" ] ;
+Item = Headers "\n" Payload ;
+Payload = { * } ;
+```
+
+- **Headers** are a single line containing a JSON object, as defined in the
+  *Headers* section. Attributes defined in the Envelope header scope the
+  contents of the Envelope and can be thought of applying to all items.
+
+- Based on the contents of the Envelope, certain attributes may be required. See
+  [Data Model](#data-model) for a specification of required attributes.
+- There can be an arbitrary number of **Items** separated by a newline. An
+  implementation should consume items until the file ends. Envelope may be
+  empty, terminating immediately after the headers.
+- Each Envelope should be terminated with a trailing newline. This newline is
+  optional. After the final newline, no whitespace is allowed.
+- The end of file (EOF) does not implicitly terminate an Envelope if more data
+  is expected.
+
+### Items
+
+Items carry actual information of an Envelope. Without items, an Envelope is
+considered *empty* and can safely be discarded.
+
+There are two generic headers for every item:
+
+`type`
+
+: **String, required.** Specifies the type of this item and its contents. Based
+  on the item type, more headers may be required. See Data Integrity for a list
+  of all item types.
+
+`length`
+
+: *int, recommended.* The length of the payload in bytes. If no `length` is
+  specified, the payload implicitly goes to the next newline. For payloads
+  containing newline characters, the `length` must be specified.
+
+{% capture __alert_content -%} 
+The payload length should be declared in all cases as this may speed up parsing
+of an Envelope considerably.
+
+If the Envelope contains a high number of very small items, omitting the length
+can be beneficial for compression. This is the case for session events.
+
+The implementor should assess this in a per-case basis and explicitly argue
+about the decision.
+{%- endcapture -%}
+{%- include components/alert.html
+  title="On omitting `length`"
+  content=__alert_content
+  level="warning"
+%}
+
+
+Notes for implementors:
+
+- Implementations **must gracefully skip** items of unknown type, along with
+  their payload.
+- Unknown attributes must be forwarded to the upstream.
+- If `length` cannot be consumed, i.e. the Envelope is EOF before the number of
+  bytes has been consumed, then the Envelope is malformed.
+- If an Item with implicit length is terminated by `\r\n`, then `\r` is
+  considered an arbitrary character not part of the newline, and thus part of
+  the payload.
+- After length-prefixed item payloads there is always a `\n` or EOF. That
```

Comment:
Suggest, "Length-prefixed payloads must terminate with `\n` or EOF. The newline is not considered part of the payload. Any other character, including whitespace, means the Envelope is malformed."